### PR TITLE
Add section `Profiling your K semantics` section in the user manual.

### LINKED ISSUE #2755: Update the USERMANUAL with info about how to do rule profiling in the LLVM backend
This could be useful for profiling the Plutus semantics
@ChristianoBraga and @gtrepta can you have a look into this?

### PR DESCRIPTION
**Pull request by ChristianoBraga**
The new section, right after `Debugging`, gives a general description of the process necessary for profiling and illustrates it with a couple of Makefile target examples from the KPlutus project.

### CODE REVIEW

File: USER_MANUAL.md [hunk 1]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+export CFLAGS="-g -O2 -fno-omit-frame-pointer"; export CXXFLAGS="-g -O2 -fno-omit-frame-pointer"
+```
+
+2. You can skip this step, but if you do, any samples within these libraries will not have correct stack trace information, which means you will likely not get a meaningful set of data that will tell you where the majority of time is really being spent. Don't worry about rebuilding literally every single dependency though. Just focus on the ones that you expect to take a non-negligible amount of runtime. You will be able to tell if you haven't done enough later, and you can go back and rebuild more.
+Once this is done, you then build K with optimizations and debug info enabled, like so: 
+```shell
+mvn package -Dproject.build.type="RelWithDebInfo"
+```
+
+3. Next, you build the semantics with optimizations and debug info enabled (i.e., kompile -O3 -ccopt -g )
```
Comment (radumereuta) [inline USER_MANUAL.md]:
```suggestion
3. Next, you build the semantics with optimizations and debug info enabled (i.e., `kompile -O3 -ccopt -g` )
```

File: USER_MANUAL.md [hunk 2]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+sets debuging and optimization flags. Then it triggers making K's
+dependencies (such as the blockchain plugin) with `k-deps`. (The need
+for `-fno-omit-frame-pointer` here is related with step 2 in _General Instructions_ above.)
+
+```makefile
+k-deps-profiling: K_MVN_ARGS += -Dhaskell.backend.skip=true                                                                                                                
+k-deps-profiling: K_BUILD_TYPE := FastBuild                                                                                                                                
+k-deps-profiling: export CMAKE_CXX_FLAGS='-O2 -DNDEBUG -fno-omit-frame-pointer'                                                                                            
+k-deps-profiling: k-deps                                                                                                                                                   
+```
```
Comment (radumereuta) [inline USER_MANUAL.md]:
```suggestion
```makefile
k-deps-profiling: K_MVN_ARGS += -Dhaskell.backend.skip=true
k-deps-profiling: K_BUILD_TYPE := FastBuild
k-deps-profiling: export CMAKE_CXX_FLAGS='-O2 -DNDEBUG -fno-omit-frame-pointer'
k-deps-profiling: k-deps
```
```

File: USER_MANUAL.md [hunk 3]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+these flags.
+
+Note that both targets require removing the frame pointer
+elimination optimization. Here for the backend itself and in
+`k-deps-profiling` for the libraries that the backend uses.
+
+```makefile
+build-llvm-profiling: KOMPILE_OPTS += -O3 -ccopt -fno-omit-frame-pointer                                                                                                   
+build-llvm-profiling: $(KPLUTUS_LIB)/$(llvm_kompiled)                                                                                                                      
+```
```
Comment (radumereuta) [inline USER_MANUAL.md]:
```suggestion
```makefile
build-llvm-profiling: KOMPILE_OPTS += -O3 -ccopt -fno-omit-frame-pointer
build-llvm-profiling: $(KPLUTUS_LIB)/$(llvm_kompiled)
```
```

**Review by radumereuta — COMMENTED** [review]
Just some minor aesthetic changes.
Suggestions from the makefile blocks removes trailing whitespace so it doesn't introduce a horizontal scroll bar in the rendered version.

I would still like it this gets one more review from @dwightguth or @ehildenb

File: USER_MANUAL.md [hunk 4]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
 
+Profiling your K semantics
+--------------------------
+
+### General instructions
+<!-- _(This is this [post](https://runtimeverification.slack.com/archives/C7E701MFG/p1578937371082200), by @dwightguth, with minor changes in prose and formatting.)_ -->
+
+1. The first thing to be aware of is in order to get meaningful data, you need to build the semantics and all of its dependencies with optimizations enabled but _without the frame pointer elimination optimization_. For example, for EVM, this means rebuilding GMP, MPFR, JEMalloc, Crypto++, SECP256K1, etc with the following `exports`.
+```shell
+export CFLAGS="-g -O2 -fno-omit-frame-pointer"; export CXXFLAGS="-g -O2 -fno-omit-frame-pointer"
```
Comment (dwightguth) [inline USER_MANUAL.md]: let's add `-DNDEBUG` but remove `-g` here.

File: USER_MANUAL.md [hunk 5]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+```
+
+2. You can skip this step, but if you do, any samples within these libraries will not have correct stack trace information, which means you will likely not get a meaningful set of data that will tell you where the majority of time is really being spent. Don't worry about rebuilding literally every single dependency though. Just focus on the ones that you expect to take a non-negligible amount of runtime. You will be able to tell if you haven't done enough later, and you can go back and rebuild more.
+Once this is done, you then build K with optimizations and debug info enabled, like so: 
+```shell
+mvn package -Dproject.build.type="RelWithDebInfo"
```
Comment (dwightguth) [inline USER_MANUAL.md]: This is out of date; I would recommend using `FastBuild` instead as `RelWithDebInfo` enables LTO which does not cooperate very well with `-fno-omit-frame-pointer`.

File: USER_MANUAL.md [hunk 6]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+```
+
+3. Next, you build the semantics with optimizations and debug info enabled (i.e., `kompile -O3 -ccopt -g` )
```
Comment (dwightguth) [inline USER_MANUAL.md]: -O3 enables at least one optimization in the llvm backend that can cause some issues in some cases. It's probably better to recommend `kompile -ccopt -O2 --iterated -ccopt -fno-omit-frame-pointer`.

File: USER_MANUAL.md [hunk 7]
```diff
@@ -3042,6 +3042,85 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+llvm-kompile-compute-loc 5868 evm-semantics/.build/defn/llvm/driver-kompiled
+```
+spits out the following text.
+```
+Line: 18529
+/home/dwightguth/evm-semantics/./.build/defn/llvm/driver.k:493:10
+```
+This is the line of `definition.kore` that the axiom appears on as well as the original location of the rule in the K semantics. You can use this information to figure out which rules and functions are causing the most time and optimize them to be more efficient.
+
+## Makefile target examples
```
Comment (dwightguth) [inline USER_MANUAL.md]: I think this section just makes the above more confusing and should probably be removed.

**Review by dwightguth — CHANGES_REQUESTED** [review]

File: USER_MANUAL.md [hunk 8]
```diff
@@ -3042,6 +3042,43 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
 -   `(gdb) break definition.kore:break -> No source file named definition.kore.`
 send `-ccopt -g` to kompile in order to generate debug info symbols.
 
+Profiling your K semantics
+--------------------------
+
+### General instructions
+<!-- _(This is this [post](https://runtimeverification.slack.com/archives/C7E701MFG/p1578937371082200), by @dwightguth, with minor changes in prose and formatting.)_ -->
```
Comment (ehildenb) [inline USER_MANUAL.md]: Remove this comment.

**Comment by ehildenb** [comment]
The rest of the file is wrapped at 80 characters I believe @ChristianoBraga . Please follow the same convention.

**Comment by dwightguth** [comment]
As a matter of fact, because of how terminals that are exactly 80 characters wide tend to work, the entire file is word wrapped such that each line (with the exception of some code fragments and urls that aren't easy to effectively wrap) is *less than* 80 characters, as in, the 79th character is the last one that is allowed to exist.

File: USER_MANUAL.md [hunk 9]
```diff
@@ -3042,6 +3042,76 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+### General instructions
+
+1. The first thing to be aware of is in order to get meaningful data,
+you need to build the semantics and all of its dependencies with
+optimizations enabled but _without the frame pointer elimination
+optimization_. For example, for EVM, this means rebuilding GMP, MPFR,
+JEMalloc, Crypto++, SECP256K1, etc with the following `exports`.
+```shell export CFLAGS="-g -O2 -fno-omit-frame-pointer"; export
+CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer" ```
+
```
Comment (ehildenb) [inline USER_MANUAL.md]: @ChristianoBraga the formatting got messed up here, I guess you called `format` to do the line wrapping.

Please make sure to review the changes yourself (self-review) in browser.

File: USER_MANUAL.md [hunk 10]
```diff
@@ -3042,6 +3042,84 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+optimizations enabled but _without the frame pointer elimination
+optimization_. For example, for EVM, this means rebuilding GMP, MPFR,
+JEMalloc, Crypto++, SECP256K1, etc with the following `exports`. 
+```shell
+  export CFLAGS="-g -O2 -fno-omit-frame-pointer" ;
+  export CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
+```
+
+You can skip this step, but if you do, any samples within these
+libraries will not have correct stack trace information, which means
```
Comment (ehildenb) [inline USER_MANUAL.md]: Look at the rest of this flie, and format this the same way as the rest.

A space before the backticks codeblock, a space after it, no spaces before each line in the code block, etc...

You should _always_ inspect the formatting of th erest of the file or project or whatever you're contributing to, it's up to you, the PR author, to make sure this is correct. Burden should not be shifted onto the reviewers.

**Comment by ChristianoBraga** [comment]
I dont’t understand why this was assigned to me in the first place.

Please reassign it to someone who is responsible for formatting the
documentation of K.

**Review by ehildenb — APPROVED** [review]

File: USER_MANUAL.md (hunk 4 above)
Comment (ehildenb) [reply USER_MANUAL.md]: @dwightguth is this suggested change supposed to be for both CFLAGS and CXXFLAGS? Currently, we are saying `-DNDEBUG` for `CXXFLAGS`, but `-g` for `CFLAGS`.

Comment (dwightguth) [reply USER_MANUAL.md]: Both

**Review by dwightguth — APPROVED** [review]

File: USER_MANUAL.md [hunk 11]
```diff
@@ -3042,6 +3042,91 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+The first thing to be aware of is in order to get meaningful data,
+you need to build the semantics and all of its dependencies with
+optimizations enabled but _without the frame pointer elimination
+optimization_. For example, for EVM, this means rebuilding GMP, MPFR,
+JEMalloc, Crypto++, SECP256K1, etc with the following `exports`. 
+
+```sh
+export CFLAGS="-g -O2 -fno-omit-frame-pointer" ;
+export CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
+```
```
Comment (ehildenb) [inline USER_MANUAL.md]:
```suggestion
```sh
export CFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
export CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
```
```

File: USER_MANUAL.md [hunk 12]
```diff
@@ -3042,6 +3042,90 @@ Using `rbreak <regex>` you can set breakpoints on multiple functions.
...
+
+The first thing to be aware of is in order to get meaningful data,
+you need to build the semantics and all of its dependencies with
+optimizations enabled but _without the frame pointer elimination
+optimization_. For example, for EVM, this means rebuilding GMP, MPFR,
+JEMalloc, Crypto++, SECP256K1, etc with the following `exports`. 
+
+```sh
+export CFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
+export CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
```
Comment (ehildenb) [inline USER_MANUAL.md:3056]:
`````suggestion
```sh
export CFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
export CXXFLAGS="-DNDEBUG -O2 -fno-omit-frame-pointer"
```
`````

**Pull request merged**